### PR TITLE
feat: added watermark, judge, legal advisor and final flag for sdo (FPLA-69)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-69

### Change description ###

added watermark to draft document
added judge and legal advisor field
added flag for draft / final version of sdo

BREAKING CHANGE TO CURRENT IMPLEMENTATION OF JUDGE AND LEGAL ADVISOR IN C6

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes
[] No
```

### CODE REVIEW

File: ccd-definition/ComplexTypes/JudgeAndLegalAdvisor.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "ElementLabel": "Judge or magistrate's title",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "JudgeAndLegalAdvisor",
+    "ListElementCode": "judgeLastName",
+    "FieldType": "Text",
+    "ElementLabel": "Last name",
+    "FieldShowCondition": "judgeTitle=\"HER_HONOUR_JUDGE\" OR judgeTitle=\"HIS_HONOUR_JUDGE\" OR judgeTitle=\"DEPUTY_DISTRICT_JUDGE\"",
```
Comment (lewisbirks) [inline ccd-definition/ComplexTypes/JudgeAndLegalAdvisor.json]: Couldn't this use `CONTAINS` as they all contain `JUDGE`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.hmcts.reform.fpl.model.common;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.hmcts.reform.fpl.enums.JudgeOrMagistrateTitle;
+
+@Data
+@Builder
+@AllArgsConstructor
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java]: Not needed when using `@Data` when all fields are final

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/DocumentReference.java [hunk 3]
```diff
@@ -6,7 +6,7 @@
 import lombok.Data;
 
 @Data
-@Builder
+@Builder(toBuilder = true)
 @AllArgsConstructor
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/DocumentReference.java]: Not needed when using `@Data` when all fields are final

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java [hunk 4]
```diff
@@ -78,37 +91,90 @@ public CaseDataExtractionService(DateFormatterService dateFormatterService,
...
+        }
+
+        return data.build();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> prepareJudgeAndLegalAdvisor(JudgeAndLegalAdvisor judgeAndLegalAdvisor) {
+        ImmutableMap.Builder map = ImmutableMap.<String, Object>builder();
+
+        if (judgeAndLegalAdvisor == null || isEmpty(judgeAndLegalAdvisor)) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java]: `IsEmpty` performs a null check and returns true if null

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/JudgeOrMagistrateTitle.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.reform.fpl.enums;
+
+public enum JudgeOrMagistrateTitle {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/JudgeOrMagistrateTitle.java]: Could use lombok here with a `@RequiredArgsContructor` and `@Getter` to reduce boiler plate

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderStatus.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.hmcts.reform.fpl.enums;
+
+public enum OrderStatus {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderStatus.java]: Could use lombok here with a `@RequiredArgsContructor` and `@Getter` to reduce boiler plate

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 7]
```diff
@@ -69,6 +71,9 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
                 caseData.getStandardDirectionOrder().getDirections());
 
             directions.forEach((key, value) -> caseDetails.getData().put(key, value));
+
+            caseDetails.getData().put("judgeAndLegalAdvisor",
+                defaultIfNull(caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor(), null));
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java]: Does this actually change anything? Could you just use `caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor()`?

File: ccd-definition/ComplexTypes/JudgeAndLegalAdvisor.json (hunk 1 above)
Comment (swalker125) [reply ccd-definition/ComplexTypes/JudgeAndLegalAdvisor.json]: I flagged this on the PR that implemented this change... it was apparently not possible 🤷‍♂

Comment (swalker125) [reply ccd-definition/ComplexTypes/JudgeAndLegalAdvisor.json]: I shall look into

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java (hunk 7 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java]: nope, removed

**Comment by Jamesnelson117** [comment]
Had a quick look at your template, you're harding court location. Do we need to update caseDataValidationService to return courtLocation? Similar to how we do on c6 c6a

File: e2e/pages/events/draftStandardDirectionsEvent.page.js [hunk 8]
```diff
@@ -15,9 +16,21 @@ module.exports = {
...
+        draft: 'No, I need to make changes',
+      },
     };
   },
 
+  async enterJudgeAndLegalAdvisor(judgeLastName, legalAdvisorName) {
+    judgeAndLegalAdvisor.selectJudgeTitle();
+    judgeAndLegalAdvisor.enterJudgeLastName(judgeLastName);
+    judgeAndLegalAdvisor.enterLegalAdvisorName(legalAdvisorName);
+    await I.retryUntilExists(() => I.click('Continue'), '#allParties');
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply e2e/pages/events/draftStandardDirectionsEvent.page.js]: Defined in `steps_file.js`. Max number is 4 it seems

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java [hunk 9]
```diff
@@ -3,7 +3,7 @@
 public enum DocmosisTemplates {
     C6("FL-PLW-HNO-ENG-00092.docx", "Notice_of_proceedings_(c6)"),
     C6A("FL-PLW-HNO-ENG-00093.docx", "Notice_of_proceedings_(c6a)"),
-    SDO("FL-PLW-STD-ENG-00099.docx", "Standard_directions_order");
+    SDO("FL-PLW-STD-ENG-00099.docx", "draft_standard_directions_order.pdf");
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java]: I've just commented the same... doing so means you don't need to pass in a the document title elsewhere 

example: 
`   DocmosisDocument document = docmosisService.generateDocmosisDocument(templateData, DocmosisTemplates.SDO);

        return uploadDocumentService.uploadPDF(userId, authorization, document.getBytes(), document.getDocumentTitle());`

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionServiceTest.java [hunk 10]
```diff
@@ -124,12 +135,19 @@ void shouldMapCompleteCaseDataForSDOTemplate() throws IOException {
...
             .dateSubmitted(LocalDate.now())
             .respondents1(createRespondents())
-            .standardDirectionOrder(createStandardDirectionOrders(TODAYS_DATE_TIME))
+            .standardDirectionOrder(createStandardDirectionOrders(TODAYS_DATE_TIME, OrderStatus.SEALED))
             .build();
 
         Map<String, Object> templateData = caseDataExtractionService
             .getStandardOrderDirectionData(caseData);
 
+        System.out.println("templateData = " + templateData);
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionServiceTest.java]: 👀

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java [hunk 11]
```diff
@@ -175,8 +246,8 @@ private String getFirstApplicantName(CaseData caseData) {
             .map(Child::getParty)
             .map(child -> ImmutableMap.of(
                 "name", child.getFirstName() + " " + child.getLastName(),
-                "gender", defaultIfNull(child.getGender(), EMPTY_STATE_PLACEHOLDER),
-                "dateOfBirth", child.getDateOfBirth() == null ? EMPTY_STATE_PLACEHOLDER :
+                "gender", defaultIfNull(child.getGender(), EMPTY_PLACEHOLDER),
+                "dateOfBirth", child.getDateOfBirth() == null ? EMPTY_PLACEHOLDER :
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java]: 👍

File: ccd-definition/CaseEventToFields/draftSDO.json [hunk 12]
```diff
@@ -177,9 +189,9 @@
     "CaseFieldID": "standardDirectionOrder",
     "PageFieldDisplayOrder": 1,
     "DisplayContext": "COMPLEX",
-    "PageID": 7,
+    "PageID": "documentReview",
     "PageLabel": " ",
-    "PageDisplayOrder": 7,
+    "PageDisplayOrder": 8,
     "PageColumnNumber": 1
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/CaseEventToFields/draftSDO.json]: you can have many columns of content on one page, we only ever make use of column 1

File: ccd-definition/AuthorisationCaseField/gatekeeper.json [hunk 13]
```diff
@@ -111,6 +111,13 @@
     "UserRole": "caseworker-publiclaw-gatekeeper",
     "CRUD": "R"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "judgeAndLegalAdvisor",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRUD"
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/AuthorisationCaseField/gatekeeper.json]: can actually be just RU :)

File: e2e/pages/events/draftStandardDirectionsEvent.page.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (swalker125) [reply e2e/pages/events/draftStandardDirectionsEvent.page.js]: added by Damian in his e2e refactoring 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 14]
```diff
@@ -115,28 +121,47 @@ public AboutToStartOrSubmitCallbackResponse handleMidEvent(
...
+        Order order = updated.getStandardDirectionOrder().toBuilder()
+            .orderDoc(DocumentReference.builder()
+                .url(document.links.self.href)
+                .binaryUrl(document.links.binary.href)
+                .filename(documentFilename(caseData.getStandardDirectionOrder().getOrderStatus()))
+                .build())
+            .build();
+
+        caseDetails.getData().put("standardDirectionOrder", order);
+        caseDetails.getData().remove("judgeAndLegalAdvisor");
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java]: have added a judgeAndLegalAdvisor.isNull() assertion

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java [hunk 15]
```diff
@@ -17,22 +18,31 @@
...
 import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang.StringUtils.defaultIfBlank;
+import static org.apache.commons.lang.StringUtils.isNotBlank;
 import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+import static org.apache.commons.lang3.ObjectUtils.isNotEmpty;
+import static uk.gov.hmcts.reform.fpl.enums.OrderStatus.SEALED;
 
 // Supports SDO case data. Tech debt ticket needed to refactor caseDataExtractionService and NoticeOfProceedingsService
+@Slf4j
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseDataExtractionService.java]: hehe 😉

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 16]
```diff
@@ -148,7 +173,13 @@ private Document getDocument(@RequestHeader("authorization") String authorizatio
...
-        return uploadDocumentService.uploadPDF(userId, authorization, document.getBytes(),
-            "draft-standard-directions-order.pdf");
+        return uploadDocumentService.uploadPDF(userId, authorization, document.getBytes(), document.getDocumentTitle());
+    }
+
+    private String documentFilename(OrderStatus status) {
+        if (status == OrderStatus.SEALED) {
+            return OrderStatus.SEALED.getDocumentTitle();
+        }
+        return OrderStatus.DRAFT.getDocumentTitle();
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java]: Method can be removed and just get the document title from where the it is called

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java]: Yep makes sense, will make the same change in my own PR

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java (hunk 16 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java]: silly me, thanks 👍

**Pull request merged**